### ISSUE TITLE
edge case parsing rules

### ISSUE THREAD

**Issue by orefalo** [lmorg/murex#703]
**Describe the bug:**

Good morning,
I believe I found a small discrepancy

```
➜  out {GREEN}{BOLD}Complete!{RESET}
  Complete!
➜  out {GREEN}{GREEN}{BOLD}Complete!{RESET}
   Complete!
```

Note how spaces are added to the beginning of the line. In fact, it looks like a space is added for each `{ANSI}`

*** Expected behavior ***

colors & styles but no space
```
➜  out {GREEN}{BOLD}Complete!{RESET}
Complete!
➜  out {GREEN}{GREEN}{BOLD}Complete!{RESET}
Complete!
```

Thank you

**Comment by lmorg** [comment]
You’ll need to wrap the string in quotes.

the issue is that {code}{code} when parsed as barewords are treated as executable blocks so different parameters. Out supports multiple parameters but concatenated them with spaces. 

It’s definitely a bug but I’m not sure how best to solve this one. On the one hand, you’d expect statements like `if {condition}{action}` to be split. But then on the other hand, `out {code}{code}` shouldn’t. However they’re both just functions with bareword parameters. So how do you know which rule to apply to what?

the only solution I can think of is to set signatures for each function so that the parser can apply different parsing rules based on what types you assign to each parameter. But that can get really messy really fast when you consider that functions can be overloaded to have any combination of parameters.

in short, I haven’t devised a great way to solve this problem yet.

**Comment by orefalo** [comment]
I understand the problem...
   Please help me here, why should `if {condition}{action}` be split in the first place. Correct me if I am wrong, but the parser should handle it with no issues, with or without spaces.

Few years ago, I [rewrote the GraphQL parser](https://github.com/orefalo/graphql-shorthand-parser/blob/master/src/graphql-shorthand.pegjs) using PEGJS. Do you have a similar grammar for the murex syntax in the code? can you point me to it?

**Comment by lmorg** [comment]
The issue is with barewords. This is why almost all languages limit barewords to a minimum, forcing strings to be quoted. The few that do support barewords to the same extent as shells have a thousand weird edge cases (like YAML).

`out "{GREEN}{BOLD}Complete!{RESET}"` parses fine because you're quoting the string. But without the quotation marks, like in your example, `{` and `}` are ambitious. Is it token or is it a string. If I parse it as a token then your example breaks. If i parse it as a string then `if {condition}{action}` breaks. Both are valid syntax from our perspective but only one can be from a grammar perspective. So how do you decide which is correct and which isn't?

The simple solution is to enforce quotation marks on strings. But that's horrible from a shell ergonomics perspective. So I need to come up with something more elaborate :(

**Comment by orefalo** [comment]
got it - indeed.
closing, thank you

**Comment by lmorg** [comment]
Have had a thought. It’s a bit of a hack but I’m pretty sure it’s the same way bash knows when to treat `function {}` as multi line vs `echo {}` - I could just make `{}` a token conditional to what the command is.

it’s a massive kludge however it is a very simple fix and could remove a lot of surprises. 

I was thinking the same kind of approach could be used for parsing variables as well. One of my pet niggles is that it’s very easy to put $variable in `set` and `foreach` (et al) and doing so would cause your script to break in unexpected ways. So having a way to tell the parser to parse $variable as a string rather than a variable under specific conditions would also help with UX. 

What’s your thoughts?

**Comment by lmorg** [comment]
Reopening because while not a bug per se, it’s definitely an issue.

**Comment by orefalo** [comment]
> So having a way to tell the parser to parse $variable as a string rather than a variable under specific conditions would also help with UX.

Definitely -> users (and CX) first -
Now this should be completely internal to murex.

**Comment by lmorg** [comment]
I have a plan for the ansi codes. The original plan was always to have another special quotation mark that would trigger it so that they could work in any command and not just with select few ones. The problem was my parser wasn’t sophisticated enough to do that. But last year I completely rewrote the parser and introduced `%()` for strings. So I now have all the pieces in place to implement ansi support. And the best thing here is that `%({ANSI}{ANSI})` would always get parsed correctly because it’s explicitly an ansi string and not a bareword.


I just need to put the code in place

**Comment by orefalo** [comment]
I am sure that `%()` has super mighty powers, but then I would rather use `""` ;-)

Anyway, I believe this comment is outdated based on the other thread. it works fine now.
